### PR TITLE
Improve youtube api calls

### PR DESCRIPTION
**Pull request by SamantazFox**
See #1981 and https://github.com/iv-org/invidious/pull/1929#discussion_r602996585 for previous discussions on the subject.

This PR is only a partial job, as some code cleaning and some other fixes are required before continuing. See https://github.com/iv-org/invidious/pull/1985#issuecomment-846607801 and https://github.com/iv-org/invidious/pull/1985#issuecomment-847027137 for further details.

### CODE REVIEW

File: src/invidious/helpers/youtube_api.cr [hunk 1]
```diff
@@ -30,9 +30,49 @@ def request_youtube_api_browse(continuation)
     "continuation": continuation,
   }
 
+  return _youtube_api_post_json("/youtubei/v1/browse", data)
+end
+
+####################################################################
+# request_youtube_api_search(search_query, params, region)
+#
+# Requests the youtubei/vi/search endpoint with the required headers
```
Comment (TiA4f8R) [inline src/invidious/helpers/youtube_api.cr:39]:
```suggestion
# Requests the youtubei/v1/search endpoint with the required headers
```

Comment (SamantazFox) [reply src/invidious/helpers/youtube_api.cr:39]: Oh, right, thanks :) I'll fix those later in a separate commit!

**Comment by unixfox** [comment]
Thank you for your work. Great improvements. That's the way to go in order to reduce the amount of reCAPTCHA and have a more stable way of fetching the data from YouTube.

**Comment by SamantazFox** [comment]
> Thank you for your work. Great improvements. That's the way to go in order to reduce the amount of reCAPTCHA and have a more stable way of fetching the data from YouTube.

No probs ^^

I'm currently facing an issue, tho: the actual way of fetching video infos (`ytInitialData` and `ytInitialPlayerResponse`) is a bit tricky because cookies are saved, too, and there are workarounds for region locked videos.

**Comment by TiA4f8R** [comment]
@SamantazFox Note also that there are two endpoints used for videos: `player` (in the `streamingData` is present for example) and `next` (for additional stats and next videos) (same body for the POST request).

**Comment by SamantazFox** [comment]
> @SamantazFox Note also that there are two endpoints used for videos: `player` (in the `streamingData` is present for example) and `next` (for additional stats and next videos) (same body for the POST request).

Thanks! For now, I'm only interested in using the `player` endpoint, as next video(s) and other stats are not used in invidious (Invidious relies on playlist data to know what to play next in the case of a playlist, and doesn't autoplay next video in other cases).

**Comment by TiA4f8R** [comment]
When I said additional stats, it's likes/dislikes for example and the next videos, which are part of the watch page of Invidious, even if they are not from a playlist.

**Comment by FireMasterK** [comment]
There's not rate limit according to my testing. (I tested 35k requests).

You may be interested in using the mobile client name and version to remove the need for signature ciphers for the API.

Another thing I noticed - Invidious has support for protobuf requests?! You can replicate the exact android app for even api stablity with that!

**Comment by SamantazFox** [comment]
> When I said additional stats, it's likes/dislikes for example and the next videos, which are part of the watch page of Invidious, even if they are not from a playlist.

Oh, okay, nice! Thanks for the clarification!


> There's not rate limit according to my testing. (I tested 35k requests).

Nifty :D
 
> You may be interested in using the mobile client name and version to remove the need for signature ciphers for the API. (`ANDROID` and `16.02.35`)

What do you mean by "signature ciphers"?
 
> Another thing I noticed - Invidious has support for protobuf requests?! You can replicate the exact android app for even api stablity with that!

Yes, thanks to [Omar Roth's `protodec`](https://github.com/omarroth/protodec)! I don't have the app, so if you can povide the different protobuf objects on Android (along with metadata, like playlist ID, videoID, cannel ID, etc...), that would be cool!

We could also randomize between the two, to reduce even more the risk of rate limiting.

**Comment by TiA4f8R** [comment]
> What do you mean by "signature ciphers"?

See https://github.com/iv-org/invidious/blob/master/src/invidious/helpers/signatures.cr. It's for the protected contents, especially music contents (YouTube Music tracks, videoclips, ...).

**Comment by SamantazFox** [comment]
> See https://github.com/iv-org/invidious/blob/master/src/invidious/helpers/signatures.cr. It's for the protected contents, especially music contents (YouTube Music tracks, videoclips, ...).

Oooh, right! From my experience with invidious (about a year or so), reading of protected videos never worked. I guess that's yet another entire subject.

Also, from my understanding, we will alway need those ciphers, as they're required to get the video stream from youtube, no matter wich API we use to get JSON: https://github.com/iv-org/invidious/blob/b7eba3bc178b76a733a142f475cdb83921d591c1/src/invidious/videos.cr#L584

**Comment by TiA4f8R** [comment]
Not really, see TeamNewPipe/NewPipeExtractor#562.

**Comment by SamantazFox** [comment]
Small update: I've been using this branch on my personnal instance for 3 weeks, and it works without issues. I'll try to continue working on that this week-end (I'm still having some server issues).

**Comment by TiA4f8R** [comment]
Nice ;)

File: src/invidious/helpers/youtube_api.cr [hunk 2]
```diff
@@ -20,12 +35,86 @@ def request_youtube_api_browse(continuation)
...
+# paramters, formatted as a base64 string.
+#
+def request_youtube_api_search(search_query : String, params : String, region = nil)
+  # JSON Request data, required by the API
+  data = {
+    "query":   URI.encode_www_form(search_query),
+    "context": {
+      "client": {
+        "hl":            "en",
+        "gl":            region || "US", # Can't be empty!
```
Comment (unixfox) [inline src/invidious/helpers/youtube_api.cr:80]: If you want to ignore the warning about the value can't be a nil you can do it this way instead:
```json
"gl": region.not_nil!
```

**Comment by unixfox** [comment]
What's left to be worked on here? In case I could help.

**Comment by TiA4f8R** [comment]
> What's left to be worked on here? In case I could help.

The use of the new internal API for the videos in Invidious (see my issue for more details).

**Comment by syeopite** [comment]
A lot of scraping needs to be replaced by this as well but that *may* be a bit out of scope for this specific PR

**Comment by SamantazFox** [comment]
> What's left to be worked on here? In case I could help.

* The community tab code needs to be fixed (See #1979) before moving it to the `browse` endpoint:
https://github.com/SamantazFox/invidious/blob/improve-youtube-api-helper/src/invidious/channels.cr#L502

* The `fetch_channel_playlists` and `get_about_info` function should use the `browse` endpoint:
https://github.com/SamantazFox/invidious/blob/improve-youtube-api-helper/src/invidious/channels.cr#L358
https://github.com/SamantazFox/invidious/blob/improve-youtube-api-helper/src/invidious/channels.cr#L777

* The video/player code needs to be documented & tested. I have no clue of what is still needed/what is obsolete, nor how the region bypass code is working. This is required if we need to use the `video` and `next` endpoints of the youtube API.

* The API helper should also return the HTTP response headers (`Cookies`, `Location`, ...) along with the rest of the data, as it is probably required for region bypass.

**Comment by SamantazFox** [comment]
> A lot of scraping needs to be replaced by this as well but that _may_ be a bit out of scope for this specific PR

Do you have examples?

File: src/invidious/helpers/youtube_api.cr (hunk 2 above)
Comment (SamantazFox) [reply src/invidious/helpers/youtube_api.cr:80]: Youtube will return a 400 error code is you don't provide valid `hl` and `gl` parameters, so `region` CAN'T be `nil`

**Comment by TiA4f8R** [comment]
> This is required if we need to use the `video` and `next` endpoints of the youtube API.

`player` and `next`, not `video` and `next` ;)

**Comment by SamantazFox** [comment]
@TiA4f8R ah, yeah, thanks ^^

**Comment by syeopite** [comment]
> > A lot of scraping needs to be replaced by this as well but that _may_ be a bit out of scope for this specific PR
> 
> Do you have examples?

[Here's one. ](https://github.com/iv-org/invidious/blob/master/src/invidious/channels.cr#L795-L808) There's a lot of areas in Invidious that scrapes Youtube's page.

**Comment by syeopite** [comment]
> The fetch_channel_playlists and get_about_info function should use the browse endpoint:

The `get_about_info` function should ideally be restricted to only the about channel tab. There's absolutely no reason on why we're making two requests every single time. The browse endpoint for each of the channel tabs has all of the data we need for the about information.

**Comment by SamantazFox** [comment]
> The `get_about_info` function should ideally be restricted to only the about channel tab. There's absolutely no reason on why we're making two requests every single time. The browse endpoint of each channel tab contains all of the channel about information we'll need.

We need that dual-request thing for exactly one reason: the browse API endpoint only takes a channel IDs, in the `UC[a-zA-Z0-9_-]{22}` form. But we still want to provide an access to `/user/somebody` to the end user. It means that we have, at some point, to find what id the channel ID of the `somebody` user.

Currently, the about page is the lightest available (less than 200kiB, iirc), so we can get the channel ID without too much latency.

**Comment by TiA4f8R** [comment]
> > The `get_about_info` function should ideally be restricted to only the about channel tab. There's absolutely no reason on why we're making two requests every single time. The browse endpoint of each channel tab contains all of the channel about information we'll need.
> 
> We need that dual-request thing for exactly one reason: the browse API endpoint only takes a channel IDs, in the `UC[a-zA-Z0-9_-]{22}` form. But we still want to provide an access to `/user/somebody` to the end user. It means that we have, at some point, to find what id the channel ID of the `somebody` user.
> 
> Currently, the about page is the lightest available (less than 200kiB, iirc), so we can get the channel ID without too much latency.

There is an endpoint called `resolve_url`, see what I did in my PR for NewPipe Extractor: https://github.com/TeamNewPipe/NewPipeExtractor/blob/e57e2df1e0b89191774def6e28ade6c4a1e39db0/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java#L85-L133

**Comment by SamantazFox** [comment]
> There is an endpoint called `resolve_url`, see what I did in my PR for NewPipe Extractor: https://github.com/TeamNewPipe/NewPipeExtractor/blob/e57e2df1e0b89191774def6e28ade6c4a1e39db0/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeChannelExtractor.java#L85-L133

You can't imagine how much I love you for that and the rest <3

**Comment by SamantazFox** [comment]
rebased on master

**Comment by TiA4f8R** [comment]
> You can't imagine how much I love you for that and the rest <3

I really appreciate, but say thank you to @FireMasterK. Without they, I would not have been able to find this.

**Comment by SamantazFox** [comment]
>     * The `fetch_channel_playlists` and `get_about_info` function should use the `browse` endpoint:
>       https://github.com/SamantazFox/invidious/blob/improve-youtube-api-helper/src/invidious/channels.cr#L358
>       https://github.com/SamantazFox/invidious/blob/improve-youtube-api-helper/src/invidious/channels.cr#L777

I thought I could do that easily, but the channel's routes code should be cleaned first.

**Comment by SamantazFox** [comment]
Oh come on.... It's just a useless comma....
https://github.com/iv-org/invidious/pull/1985/checks?check_run_id=2656341135

**Review by saltycrys — APPROVED** [review]
Great job!

**Comment by unixfox** [comment]
I'm a bit lost on this PR, is all the work already done? If so then I would be very happy to test this PR.

**Comment by SamantazFox** [comment]
> I'm a bit lost on this PR, is all the work already done?

No, there is still stuff to do. As explained [here](https://github.com/iv-org/invidious/pull/1985#issuecomment-846607801), there are some other things that need to be fixed/sorted out first.

TL;DR: I don't properly understand the region bypass code, the channel community tab need to be fixed first, the channel routes need to be moved out of `invidious.cr` and cleaned up.

> Like are we still parsing the HTML of the [www.youtube.com](http://www.youtube.com) webpage?

Yes. [here](https://github.com/SamantazFox/invidious/blob/b7fe212a184b5af1ccb9315e106c0ecb2f150590/src/invidious/channels.cr#L375), [here](https://github.com/SamantazFox/invidious/blob/b7fe212a184b5af1ccb9315e106c0ecb2f150590/src/invidious/channels.cr#L512-L516), [here](https://github.com/SamantazFox/invidious/blob/b7fe212a184b5af1ccb9315e106c0ecb2f150590/src/invidious/channels.cr#L791-L805] and probably in many other places....

**Comment by unixfox** [comment]
Ok so we can keep this PR as a draft until we use the majority of the youtube mobile API?

**Comment by SamantazFox** [comment]
> Ok so we can keep this PR as a draft until we use the majority of the youtube mobile API?

I'm not sure to understand what you mean by "use the majority of the YT mobile API"?

**Comment by unixfox** [comment]
> > Ok so we can keep this PR as a draft until we use the majority of the youtube mobile API?
> 
> I'm not sure to understand what you mean by "use the majority of the YT mobile API"?

Well I meant that until we don't stop scraping the HTML of the YouTube webpages for most of the logic in Invidious we can set this PR as a draft because like you said in https://github.com/iv-org/invidious/pull/1985#issuecomment-847472977 there are still a lot of code that need to be converted to use the YouTube mobile API.

**Comment by SamantazFox** [comment]
> Well I meant that until we don't stop scraping the HTML of the YouTube webpages for most of the logic in Invidious we can set this PR as a draft because like you said in [#1985 (comment)](https://github.com/iv-org/invidious/pull/1985#issuecomment-847472977) there are still a lot of code that need to be converted to use the YouTube mobile API.

Ah, ok. I'll see what I can do for the player, but this PR will need to be merged before I can do anything about the channel's ["about" tab](https://github.com/SamantazFox/invidious/blob/b7fe212a184b5af1ccb9315e106c0ecb2f150590/src/invidious/channels.cr#L777).

The `get_channel_about` function is currently used in *all* channel-related routes ([here](https://github.com/SamantazFox/invidious/blob/b7fe212a184b5af1ccb9315e106c0ecb2f150590/src/invidious.cr#L1260), [here](https://github.com/SamantazFox/invidious/blob/b7fe212a184b5af1ccb9315e106c0ecb2f150590/src/invidious.cr#L1672), [here](https://github.com/SamantazFox/invidious/blob/b7fe212a184b5af1ccb9315e106c0ecb2f150590/src/invidious.cr#L1739), ...) as a way to get the channel's UCID from a user-name.

I'd prefer to not touch that atm, and make a separate PR that will clean the code, move it to `src/invidious/routes`, and properly handle the different URL forms (i.e `/c/<ucid>`, `/channel/<ucid>`, `/u/<username>`, `/user/<username>`, and many more) by using the `resolve_url` API endpoint, as pointed out by @TiA4f8R here: https://github.com/iv-org/invidious/pull/1985#issuecomment-846944444.

**Comment by unixfox** [comment]
PR live in testing on https://yewtu.be

**Comment by SamantazFox** [comment]
> PR live in testing on https://yewtu.be

@unixfox How is it going?

**Comment by unixfox** [comment]
> > PR live in testing on [yewtu.be](https://yewtu.be)
> 
> @unixfox How is it going?

No issues on my side, it's rock solid.
